### PR TITLE
[KYUUBI #544] Deregister engine when meeting specified exception class

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Involve `kyuubi.engine.deregister.exception.classes` to deregister engine when job failed because of specified exception class

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala [hunk 1]
```diff
@@ -17,13 +17,16 @@
...
-import org.apache.spark.scheduler.{SparkListener, SparkListenerApplicationEnd}
+import org.apache.spark.scheduler.{JobFailed, SparkListener, SparkListenerApplicationEnd, SparkListenerJobEnd}
 
 import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf.ENGINE_DEREGISTER_EXCEPTION_CLASSES
 import org.apache.kyuubi.service.{Serverable, ServiceState}
 
 class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logging {
 
+  val deRegisterExceptionCLs = server.getConf.get(ENGINE_DEREGISTER_EXCEPTION_CLASSES)
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:28]: `CLs` or `Cls`?

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:28]: yeah, I wonder that Cls means a singleton class, So, I try to use CLs.

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:28]: deregisterExceptions?

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:28]: but there are exception classes, exception messages ...

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala [hunk 2]
```diff
@@ -34,4 +37,15 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
...
   }
 
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+   jobEnd match {
+     case JobFailed(e) if deRegisterExceptionCLs.exists(_.equals(e.getClass.getCanonicalName)) =>
+       error(
+         s"""
+           | Job failed exception class is in the set of ${ENGINE_DEREGISTER_EXCEPTION_CLASSES.key},
+           | stopping the engine.""".stripMargin, e)
+       server.stop()
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:47]: server.discoveryService.asInstanceOf[...].stop()

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:28]: deregisterMessages deregisterStacktraces

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala [hunk 3]
```diff
@@ -34,4 +37,15 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
     }
   }
 
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+   jobEnd match {
+     case JobFailed(e) if deRegisterExceptionCLs.exists(_.equals(e.getClass.getCanonicalName)) =>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngineListener.scala:42]: e nullable?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 4]
```diff
@@ -15,15 +15,19 @@
  * limitations under the License.
  */
 
-package org.apache.kyuubi.engine.spark
+package org.apache.spark.kyuubi
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:18]: Add some comment here

**Review by yaooqinn — APPROVED** [review]

**Review by pan3793 — APPROVED** [review]

**Pull request closed without merging**